sakura_core/doc/layout/CLayoutMgr_DoLayout.cpp

```diff
@@ -340,23 +351,97 @@ void CLayoutMgr::_DoLayout(bool bBlockingHook)
 {
 	MY_RUNNINGTIMER( cRunningTimer, L"CLayoutMgr::_DoLayout" );
 
-	/*	表示上のX位置
-		2004.03.28 Moca nPosXはインデント幅を含むように変更(TAB位置調整のため)
-	*/
-	const int nAllLineNum = m_pcDocLineMgr->GetLineCount();
+	_Empty();
+	Init();
+
+	const CLogicInt nAllLineCount = m_pcDocLineMgr->GetLineCount();
+
+	int nWorkerThreadCount = 0;
+	if (nAllLineCount < 1000) {
+		// 行数が多くなければマルチスレッド処理するまでもない
+		nWorkerThreadCount = 0;
+	} else if (CColorStrategyPool::getInstance()->HasRangeBasedColorStrategies()) {
+		// 行をまたぐ可能性のある色分け (例：/*...*/) が有効の場合、途中で処理単位が
+		// 分割されてしまうと色分けがおかしくなってしまうためやむなくマルチスレッド処理の対象外とする
+		nWorkerThreadCount = 0;
+	} else {
+		nWorkerThreadCount = (std::max)(1, (int)std::thread::hardware_concurrency()) - 1;
+	}
+
+	// 末尾側の行から順番にワーカースレッドに割り当てていき最後 (先頭行～) はメインスレッドで実行
+	// nWorkerThreadCount = 0 の場合は全行をメインスレッドが実行
+	//
+	// 1000行のテキストを10スレッドで処理する場合のイメージ：
+	// 行番号
+	//    1 ┐
+	//  ... ├ メインスレッドで処理 (i = 0)
+	//  100 ┘
+	//  101 ┐
+	//  ... ├ ワーカースレッド1で処理 (i = 1)
+	//  200 ┘
+	//  ... ...
+	//  901 ┐
+	//  ... ├ ワーカースレッド9で処理 (i = 9) ↑こっちからループ
+	// 1000 ┘
+
+	std::vector<std::thread> vecWorkerThreads;
+	std::vector<CLayoutMgr> vecWorkerLayoutMgrs(nWorkerThreadCount);
+
+	std::atomic<bool> bCanceled = false;
+	auto pbCanceled = bBlockingHook ? &bCanceled : nullptr;
+
+	int nLineIndexEnd = nAllLineCount;
+	CDocLine* pDocLineEnd = nullptr;
+	for (int i = nWorkerThreadCount; i >= 0; i--) {
+		const int nLineIndexBegin = (int)((double)nAllLineCount / (nWorkerThreadCount + 1) * i);
```

Comment:
(sonarcloud指摘) 必ずしも対応不要。

```suggestion
		const auto nLineIndexBegin = (int)((double)nAllLineCount / (nWorkerThreadCount + 1) * i);
```